Source/modules/webcl/WebCLCommandQueue.cpp

```diff
@@ -0,0 +1,1301 @@
+#include "modules/webcl/WebCLImageDescriptor.h"
+#include "modules/webcl/WebCLOpenCL.h"
+
+#include <wtf/MainThread.h>
+
+namespace blink {
+
+WebCLCommandQueue::~WebCLCommandQueue()
+{
+    release();
+    ASSERT(!m_clCommandQueue);
+}
+
+PassRefPtr<WebCLCommandQueue> WebCLCommandQueue::create(cl_command_queue commandQueue, WebCLContext* context, WebCLDevice* device)
+{
+    return adoptRef(new WebCLCommandQueue(commandQueue, context, device));
+}
+
+ScriptValue WebCLCommandQueue::getInfo(ScriptState* scriptState, int paramName, ExceptionState& es)
+{
+    v8::Handle<v8::Object> creationContext = scriptState->context()->Global();
+    v8::Isolate* isolate = scriptState->isolate();
+
+    if (isReleased()) {
+        es.throwWebCLException(WebCLException::INVALID_COMMAND_QUEUE, WebCLException::invalidCommandQueueMessage);
+        return ScriptValue(scriptState, v8::Null(isolate));
+    }
+
+    cl_int err = CL_SUCCESS;
+    cl_command_queue_properties queueProperties = 0;
+    switch(paramName) {
+    case CL_QUEUE_CONTEXT:
+        return ScriptValue(scriptState, toV8(context(), creationContext, isolate));
+        break;
+    case CL_QUEUE_DEVICE:
+        return ScriptValue(scriptState, toV8(m_device, creationContext, isolate));
+        break;
+    case CL_QUEUE_PROPERTIES:
+        err = clGetCommandQueueInfo(m_clCommandQueue, CL_QUEUE_PROPERTIES, sizeof(cl_command_queue_properties), &queueProperties, nullptr);
+        if (err == CL_SUCCESS)
+            return ScriptValue(scriptState, v8::Integer::NewFromUnsigned(isolate, static_cast<unsigned>(queueProperties)));
+        break;
+    default:
+        es.throwWebCLException(WebCLException::INVALID_VALUE, WebCLException::invalidValueMessage);
+        return ScriptValue(scriptState, v8::Null(isolate));
+    }
+
+    WebCLException::throwException(err, es);
+    return ScriptValue(scriptState, v8::Null(isolate));
+}
+
+unsigned WebCLCommandQueue::getProperties()
+{
+    cl_command_queue_properties queueProperties;
+    cl_int err = clGetCommandQueueInfo(m_clCommandQueue, CL_QUEUE_PROPERTIES, sizeof(cl_command_queue_properties), &queueProperties, nullptr);
+    if (err == CL_SUCCESS)
+        return static_cast<unsigned>(queueProperties);
+
+    return 0;
+}
+
+void WebCLCommandQueue::finish(WebCLCallback* whenFinished, ExceptionState& es)
+{
+    if (isReleased() || m_callback) {
+        es.throwWebCLException(WebCLException::INVALID_COMMAND_QUEUE, WebCLException::invalidCommandQueueMessage);
+        return;
+    }
+
+    if (whenFinished) {
+        m_callback = whenFinished;
+        finishImpl(true);
+    } else {
+        finishImpl(false);
+    }
+}
+
+void WebCLCommandQueue::finishImpl(bool async)
+{
+    if (async) {
+        cl_int err = clEnqueueMarkerWithWaitList(m_clCommandQueue, 0, nullptr, m_eventForCallback);
+        if (err != CL_SUCCESS || !m_eventForCallback)
+            return;
+        clSetEventCallback(*m_eventForCallback, CL_COMPLETE, &callbackProxy, this);
```

Comment:
I think all enqueuOOO method should emit exception after this call. I guess the spec also mentions about it. WDYT?